Add check for non-identifier characters in schema names (#691)
open-telemetry/opentelemetry-configuration: CONTRIBUTING.md

[enum](https://json-schema.org/understanding-json-schema/reference/enum) values should be lower [snake case](https://en.wikipedia.org/wiki/Snake_case).

> ### Property name and enum value character set

Reviewer: By convention, type names are CamelCase a-zA-Z0-9 today as well. Should probably formalize this and add verification for it as well. I'm happy to address separately, but you could also do it in this PR if you're interested. 

Oddly, we added a schema modeling rule indicating that [property names are lower_snake_case](https://github.com/open-telemetry/opentelemetry-configuration/blob/f1602adfdda3daa5ca978f131743691b73240133/CONTRIBUTING.md?plain=1#L51-L54), but never codified the convention for types.

Author: Good catch, added it.

[enum](https://json-schema.org/understanding-json-schema/reference/enum) values should be lower [snake case](https://en.wikipedia.org/wiki/Snake_case).

### Property name and enum value character set

> Property names and enum values must match `^[A-Za-z_][A-Za-z0-9_]*$`: ASCII letters, digits, and underscores, not starting with a digit. These names become identifiers in the languages that generate code from the schema, so restricting them to this portable set means no code generator has to sanitize or translate a name (see [#690](https://github.com/open-telemetry/opentelemetry-configuration/issues/690)). The one exception is the intentional `/development` stability suffix (e.g. `detection/development`): a name is accepted when the part before that suffix is a valid identifier. This rule is enforced by `make compile-schema`, which fails the build if you add a property or enum value with an invalid identifier.

Reviewer: ```suggestion
Property names and enum values must match `^[A-Za-z_][A-Za-z0-9_]*$`. This facilitates usage as identifiers in generated code. Experimental properties are an exception, and are [denoted with a `/(development|alpha|beta)` suffix](https://github.com/open-telemetry/opentelemetry-configuration/blob/main/VERSIONING.md#experimental-features) (e.g. `detection/development`).
```

### Type name case

> Type names (the top-level types defined under `$defs`, e.g. `BatchSpanProcessor`) should be [PascalCase](https://en.wikipedia.org/wiki/Camel_case): they must match `^[A-Z][A-Za-z0-9]*$`, a leading uppercase ASCII letter followed by ASCII letters and digits. Like property names and enum values, type names become identifiers in the languages that generate code from the schema. Experimental types are denoted with an `Experimental` prefix rather than a `/development` suffix. This rule is enforced by `make compile-schema`, which fails the build if you add a type with a name that is not PascalCase.

Reviewer: Keep it terse, link to VERSIONING.md as the source of truth.

Update the [relevant section in CONTRIBUTING.md](https://github.com/open-telemetry/opentelemetry-configuration/blob/main/CONTRIBUTING.md#schema-validation) to indicate that this verification occurs in `make compile-schema`. 

```suggestion
Type names (the top-level types defined under `$defs`, e.g. `BatchSpanProcessor`) should be [PascalCase](https://en.wikipedia.org/wiki/Camel_case): they must match `^[A-Z][A-Za-z0-9]*$`. This facilitates usage as identifiers in generated code. Experimental types are [denoted with an `Experimental` prefix](https://github.com/open-telemetry/opentelemetry-configuration/blob/main/VERSIONING.md#experimental-features).
```
